snuba/datasets/events_processor_base.py

```diff
@@ -0,0 +1,358 @@
+from abc import ABC, abstractmethod
+from datetime import datetime
+import logging
+from typing import Any, Mapping, MutableMapping, Optional, Sequence
+
+from snuba import settings
+from snuba.clickhouse.columns import ColumnSet
+from snuba.consumer import KafkaMessageMetadata
+from snuba.datasets.events_format import (
+    enforce_retention,
+    extract_project_id,
+    extract_extra_contexts,
+    extract_extra_tags,
+    flatten_nested_field,
+    EventTooOld,
+)
+from snuba.processor import (
+    _as_dict_safe,
+    _boolify,
+    _collapse_uint32,
+    _ensure_valid_date,
+    _hashify,
+    _unicodify,
+    InvalidMessageType,
+    InvalidMessageVersion,
+    MessageProcessor,
+    ProcessorAction,
+    ProcessedMessage,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class EventsProcessorBase(MessageProcessor, ABC):
+    """
+    Base class for events and errors processors.
+    """
+
+    def __init__(self, promoted_tag_columns: ColumnSet):
+        self._promoted_tag_columns = promoted_tag_columns
+
+    @abstractmethod
+    def _should_process(self, event: Mapping[str, Any]) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _extract_event_id(
+        self, output: MutableMapping[str, Any], event: Mapping[str, Any],
+    ) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def extract_custom(
```

Comment:
I intentionally keep these methods public instead of protected and did not reorganize the responsibilities too much because the unit test (instead of testing the interface, tests the behavior of these internal methods one by one). Fixing this in this PR would have made the unit test for the processor irrelevant while it is the most useful way to ensure this change is correct.